crates/boojum/src/gadgets/tower_extension/fq2.rs

```diff
@@ -0,0 +1,676 @@
+    #[must_use]
+    pub fn negated<CS>(&mut self, cs: &mut CS) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let c0 = self.c0.negated(cs);
+        let c1 = self.c1.negated(cs);
+        Self::new(c0, c1)
+    }
+
+    /// Conjugates the element `c=c0+c1*u` by computing `c=c0-c1*u`.
+    #[must_use]
+    pub fn conjugate<CS>(&mut self, cs: &mut CS) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let c1 = self.c1.negated(cs);
+        Self::new(self.c0.clone(), c1)
+    }
+
+    /// Subtracts two elements of `Fq2` by subtracting their components elementwise.
+    #[must_use]
+    pub fn sub<CS>(&mut self, cs: &mut CS, other: &mut Self) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let c0 = self.c0.sub(cs, &mut other.c0);
+        let c1 = self.c1.sub(cs, &mut other.c1);
+        Self::new(c0, c1)
+    }
+
+    /// Multiply the element `a=a0+a1*u` by the element `b=b0+b1*u` using the Karatsuba method.
+    #[must_use]
+    pub fn mul<CS>(&mut self, cs: &mut CS, other: &mut Self) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        // v0 <- a0*b0, v1 <- a1*b1
+        let mut v0 = self.c0.mul(cs, &mut other.c0);
+        let mut v1 = self.c1.mul(cs, &mut other.c1);
+
+        // c0 <- v0 + beta*v1
+        let c0 = v0.sub(cs, &mut v1);
+
+        // c1 <- (a0 + a1)(b0 + b1) - v0 - v1
+        let mut a0_plus_a1 = self.c0.add(cs, &mut self.c1);
+        let mut b0_plus_b1 = other.c0.add(cs, &mut other.c1);
+        let mut c1 = a0_plus_a1.mul(cs, &mut b0_plus_b1);
+        let mut c1 = c1.sub(cs, &mut v0);
+        let c1 = c1.sub(cs, &mut v1);
+
+        Self::new(c0, c1)
+    }
+
+    /// Square the element `a=a0+a1*u` by using the Karatsuba method.
+    #[must_use]
+    pub fn square<CS>(&mut self, cs: &mut CS) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        // v0 <- a0^2, v1 <- a1^2
+        let mut v0 = self.c0.square(cs);
+        let mut v1 = self.c1.square(cs);
+
+        // c0 <- v0 + beta*v1
+        let c0 = v0.sub(cs, &mut v1);
+
+        // c1 <- (a0 + a1)^2 - v0 - v1
+        let mut a0_plus_a1 = self.c0.add(cs, &mut self.c1);
+        let mut c1 = a0_plus_a1.square(cs);
+        let mut c1 = c1.sub(cs, &mut v0);
+        let c1 = c1.sub(cs, &mut v1);
+
+        Self::new(c0, c1)
+    }
+
+    /// Multiply the element `a=a0+a1*u` by the element in the base field `Fq`.
+    #[must_use]
+    pub fn mul_c0<CS>(&mut self, cs: &mut CS, c0: &mut NN) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        // a*f = (a0 + a1*u)*f = (a0*f) + (a1*f)*u
+        let new_c0 = self.c0.mul(cs, c0);
+        let new_c1 = self.c1.mul(cs, c0);
+        Self::new(new_c0, new_c1)
+    }
+
+    /// Finds the inverse of the element `a=a0+a1*u` in the extension field `Fq2`.
+    #[must_use]
+    pub fn inverse<CS>(&mut self, cs: &mut CS) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
```

Comment:
I think this version will be cheaper if you compute inv like a witness and prove a*a_inv = 1

Reply:
AFAIK, inversion in torus is a fairly cheap operation, since it's additive negation `−g` in `Fq6`​ maps to the multiplicative inverse of the corresponding element in Fq12. So, since we make inner negation there is no optimizational trick like this one.